Add Episode 3 Lesson Text (#46)
czi-catalystproject/hub-champion-training: episodes/server_resources.md

(episode-server)=
# Selecting the optimal server resources for your computational work responsibly

teaching: 15
exercises: 15

:::::{card} 
OVERVIEW
^^^

::::{grid}
:gutter: 2

:::{grid-item}
Questions

- What's really happening in the cloud when I run a server?
- How many CPU cores and RAM do I need?
- How can I reinstate the default environment of a user's server?
- How can I gain control of user's server for troubleshooting purposes?

:::

:::{grid-item}
Objectives

- Select the most appropriate server resources for computational work
- Stop and restart a user's server
- Support a hub user by gaining control of their server

:::

:::::

## JupyterHub server options

> A hub provides interactive computing in the cloud to support research and education workflows for your community. Computational resources, such as the number of {ref}`CPU<glossary:cpu>` cores and {ref}`RAM<glossary:ram>`, can be provisioned on-demand and therefore allows users to scale up or scale down their work as needed.

Reviewer: ```suggestion
A hub provides interactive computing in the cloud to support research and education workflows for your community. Computational resources, such as the number of {ref}`CPU<glossary:cpu>` cores and {ref}`RAM<glossary:ram>`, can be provisioned on-demand to allow users to scale up or scale down their work as needed.
```

> Once a user is logged into the hub, they can launch a server configured with a number of {ref}`CPU<glossary:cpu>` cores and {ref}`RAM<glossary:ram>`.

Reviewer: Are the glossary references necessary a second time?

Author: Ah yes, that's me being accidentally over-zealous.

```suggestion
Once a user is logged into the hub, they can launch a server configured with a number of CPU cores and RAM`.
```

:width: 100%
:alt: Screenshot showing the File > Hub Control Panel menu option in the JupyterLab user interface.

:::

## Managing another user's server as a Hub Champion

Hub Champions have access to an Admin panel from the Hub Control Panel (or the URL `https://<your-hub-url>.2i2c.cloud/hub/admin`).

:::{image} ../media/episodes/server_resources/manage-1.png
:width: 100%
:alt: Screenshot showing the Admin panel available to Hub Champions.

:::

::::{caution}
:::{epigraph}
With great power comes great responsibility.

-- Spiderman
:::
Performing administrative actions as a Hub Champion should be done with great care. Respect the privacy of others and think before you act.

::::

### Stop and restart a user's server

A user's server may sometimes become unresponsive, e.g. issues stemming from self-installed software via `pip` and/or `conda`. Hub Champions can restore user servers to their default environment by stopping and restarting their server by clicking the ![Stop Server](../media/episodes/server_resources/manage-2.png) followed by the ![Start Server](../media/episodes/server_resources/manage-3.png) button next  to the user's name from the Admin panel.

:::{note}

> Self-installed software packages on the hub are *temporary* and are removed each time a server is shut down to maintain a clean default software environment. Software is persisted between sessions by {ref}`episode-images`, which is covered in the next episode.

Reviewer: "Persisted" feels awkward here. Is this a common usage of the verb that I am unfamiliar with?

Author: That's fair – I can change this to

```suggestion
Self-installed software packages on the hub are *temporary* and are removed each time a server is shut down to maintain a clean default software environment. Software is provided between sessions by {ref}`episode-images`, which is covered in the next episode.
```

:::

### Access a user's server

> Hub Champions's can gain control of a user's server that is currently running by clicking the ![Access Server](../media/episodes/server_resources/manage-4.png) button next to the user's name from the Admin panel.

Reviewer: ```suggestion
Hub Champions can gain control of a user's server that is currently running by clicking the ![Access Server](../media/episodes/server_resources/manage-4.png) button next to the user's name from the Admin panel.
```

> From here a Hub Champion can navigate the server *as if they are the user themselves*, which is useful for reproducing and debugging issues.

Reviewer: In the longer term if such a thing is possible, it would be cool to be able to provide some server environment that is broken in a predictable way, and ask trainee hub administrators to debug the problem as an exercise.

Author: That's a super nice suggestion – it would involve setting up a number of "broken" servers equal to the number of participants but I think it's possible. I've captured this idea in issue https://github.com/czi-catalystproject/hub-champion-training/issues/61

::::{dropdown} Exercise: User server management
:open:

Are the following statements *True* or *False*?

a. Closing the browser window is enough to stop a server.

b. Stopping a server when a user is finished with their work reduces costs and carbon emissions.

c. Stopping and restarting a server removes `pip` and `conda` user-installed packages and restores a clean state. 

d. A user server will continue to run indefinitely unless stopped.

:::{dropdown} Solution
a. *False* – the server will continue to run unless the *Stop my Server* button is clicked or will be automatically culled after 1 hour of inactivity by default.  

b. *True* – running a server utilises computational resources that are billed by the cloud provider and consumes energy.  

c. *True* – sometimes user-installed packages can introduce software dependency conflicts that can be removed by restarting the server.  

> d. *False* – servers are culled after 1 hour of inactivity by default. If a user requires a long-running process exceeding this time period, then they should request this through their Hub Champion.

Reviewer: ```suggestion
d. *False* – servers are culled after 1 hour of inactivity by default. If a user requires a long-running process exceeding this time period, then they should request an extension to this limit through their Hub Champion.
```
